File: aoc2025/day11/part2.c
#include <assert.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

typedef struct Edge {
    uint16_t dest;
} Edge;

typedef struct Node {
    Edge edges[24];
    uint8_t num_edges;
} Node;

#define MAX_NODE 17578
#define YOU 17576
#define OUT 17577

uint16_t idx(char name[3]) {
    uint16_t index = 0;
    if (!strncmp("you", name, 3)) {
        index = YOU;
    } else if (!strncmp("out", name, 3)) {
        index = OUT;
    } else {
        index = ((name[0] - 'a') * 26 * 26) + ((name[1] - 'a') * 26) +
                (name[2] - 'a');
    }

    return index;
}

Node graph[17578] = {0};

size_t dfs(uint16_t n, uint16_t end, size_t *path_counts) {
    if (n == end) {
        return 1;
    }

    if (path_counts[n] == (size_t)-1) {
        // Not yet explored
        size_t count = 0;
        for (size_t i = 0; i < graph[n].num_edges; i++) {
            count += dfs(graph[n].edges[i].dest, end, path_counts);
        }

        path_counts[n] = count;
        return count;
    } else {
        // Explored, return previous count
        return path_counts[n];
    }
}

int main() {
    char buf[1024];
    while (fgets(buf, sizeof(buf), stdin)) {
        char name[3];
        memcpy(name, buf, 3);
        uint16_t index = idx(name);

        char *ptr = strtok(buf + 5, " \n");
        do {
            memcpy(name, ptr, 3);
            uint16_t edge_index = idx(name);

            graph[index].edges[graph[index].num_edges++].dest = edge_index;
        } while ((ptr = strtok(NULL, " \n")));
    }

    size_t svr = idx("svr");
    size_t fft = idx("fft");
    size_t dac = idx("dac");

    size_t sum = 0;

    size_t path_counts[MAX_NODE] = {0};
    memset(path_counts, -1, MAX_NODE * sizeof(size_t));
    // svr -> fft -> dac -> out
    size_t first = dfs(svr, fft, path_counts);
    memset(path_counts, -1, MAX_NODE * sizeof(size_t));
    size_t second = dfs(fft, dac, path_counts);
    memset(path_counts, -1, MAX_NODE * sizeof(size_t));
    size_t third = dfs(dac, OUT, path_counts);

    sum += first * second * third;
    printf("Total: (%zu * %zu * %zu) %zu\n", first, second, third,
           first * second * third);

    memset(path_counts, -1, MAX_NODE * sizeof(size_t));
    // svr -> dac -> fft -> out
    first = dfs(svr, dac, path_counts);
    memset(path_counts, -1, MAX_NODE * sizeof(size_t));
    second = dfs(dac, fft, path_counts);
    memset(path_counts, -1, MAX_NODE * sizeof(size_t));
    third = dfs(fft, OUT, path_counts);

    sum += first * second * third;
    printf("Total: (%zu * %zu * %zu) %zu\n", first, second, third,
           first * second * third);

    printf("Sum: %zu\n", sum);
}

File: aoc2025/day11/part2.zig
const std = @import("std");

const file = @embedFile(".input.txt");
const nodes = 26 * 26 * 26;

const Node = struct {
    edges: [24]u32,
    num_edges: u32,
};

fn idx(name: []const u8) u32 {
    const one: u32 = name[0] - 'a';
    const two: u32 = name[1] - 'a';
    const thr: u32 = name[2] - 'a';
    return (one * 26 * 26) + (two * 26) + thr;
}

fn dfs(graph: []Node, current: u32, end: u32, cache: []usize) usize {
    if (current == end) {
        return 1;
    }

    if (cache[current] != std.math.maxInt(usize)) {
        return cache[current];
    }

    var count: usize = 0;
    for (graph[current].edges[0..graph[current].num_edges]) |n| {
        count += dfs(graph, n, end, cache);
    }

    cache[current] = count;
    return count;
}

pub fn main() !void {
    var graph: [nodes]Node = undefined;
    @memset(&graph, .{ .edges = undefined, .num_edges = 0 });

    var file_iter = std.mem.tokenizeScalar(u8, file, '\n');
    while (file_iter.next()) |line| {
        var line_iter = std.mem.tokenizeAny(u8, line, ": ");
        const node_index = idx(line_iter.next().?);

        graph[node_index].num_edges = 0;
        while (line_iter.next()) |edge| {
            const edge_index = idx(edge);
            graph[node_index].edges[graph[node_index].num_edges] = edge_index;
            graph[node_index].num_edges += 1;
        }
    }

    var cache = [_]usize{std.math.maxInt(usize)} ** nodes;

    const svr_to_dac = dfs(&graph, idx("svr"), idx("dac"), &cache);
    @memset(&cache, std.math.maxInt(usize));
    const dac_to_fft = dfs(&graph, idx("dac"), idx("fft"), &cache);
    @memset(&cache, std.math.maxInt(usize));
    const fft_to_out = dfs(&graph, idx("fft"), idx("out"), &cache);
    @memset(&cache, std.math.maxInt(usize));

    const svr_to_fft = dfs(&graph, idx("svr"), idx("fft"), &cache);
    @memset(&cache, std.math.maxInt(usize));
    const fft_to_dac = dfs(&graph, idx("fft"), idx("dac"), &cache);
    @memset(&cache, std.math.maxInt(usize));
    const dac_to_out = dfs(&graph, idx("dac"), idx("out"), &cache);
    @memset(&cache, std.math.maxInt(usize));

    std.log.debug("{}", .{svr_to_dac * dac_to_fft * fft_to_out + svr_to_fft * fft_to_dac * dac_to_out});
}
